src/Main.elm

```diff
@@ -1,10 +1,418 @@
+            ( turnInitialising { state | folders = Just listOfFolders }
+            , Cmd.none
+            )
+
+        ( LoadAudienceFoldersDone _, _ ) ->
+            ( model, Cmd.none )
+
+        ( LoadAudiencesDone (Err err), _ ) ->
+            ( Failed err, Cmd.none )
+
+        ( LoadAudiencesDone (Ok listOfAudiences), Initialising state ) ->
+            ( turnInitialising { state | audiences = Just listOfAudiences }
+            , Cmd.none
+            )
+
+        ( LoadAudiencesDone _, _ ) ->
+            ( model, Cmd.none )
+
+        ( GoDown childFolderID, Succeed store state ) ->
+            ( Succeed store { state | location = Folder childFolderID }
+            , Cmd.none
+            )
+
+        ( GoDown _, _ ) ->
+            ( model, Cmd.none )
+
+        ( GoUp, Succeed store state ) ->
+            ( case state.location of
+                Root ->
+                    model
+
+                Folder folderID ->
+                    let
+                        nextLocation =
+                            Store.getFolderByID folderID store
+                                |> Maybe.andThen .parent
+                                |> Maybe.map Folder
+                                |> Maybe.withDefault Root
+                    in
+                    Succeed store { state | location = nextLocation }
+            , Cmd.none
+            )
+
+        ( GoUp, _ ) ->
+            ( model, Cmd.none )
+
+        ( SetFilterBy nextFilter, Succeed store state ) ->
+            ( Succeed store
+                { state
+                    | location = Root
+                    , filterBy = nextFilter
+                }
+            , Cmd.none
+            )
+
+        ( SetFilterBy _, _ ) ->
+            ( model, Cmd.none )
+
+
+
+-------------
+-- V I E W --
+-------------
+
+
+stylesButton : List (Element.Attribute msg)
+stylesButton =
+    [ Element.width Element.fill
+    , Element.padding 20
+    ]
+
+
+viewFolder : AudienceFolder -> Element Msg
+viewFolder folder =
+    button
+        (Element.Background.color (Element.rgb255 26 113 153)
+            :: Element.Font.color (Element.rgb255 255 255 255)
+            :: stylesButton
+        )
+        { onPress = Just (GoDown folder.id)
+        , label = paragraph [] [ text ("→ " ++ folder.name) ]
+        }
+
+
+viewAudience : Audience -> Element msg
+viewAudience audience =
+    paragraph
+        (Element.Background.color (Element.rgb255 24 151 205)
+            :: Element.Font.color (Element.rgb255 255 255 255)
+            :: stylesButton
+        )
+        [ text audience.name ]
+
+
+viewGoUp : AudienceFolder -> Element Msg
+viewGoUp parentFolder =
+    button
+        (Element.Border.solid
+            :: Element.Border.width 2
+            :: Element.Border.color (Element.rgb255 26 113 153)
+            :: Element.Font.color (Element.rgb255 26 113 153)
+            :: stylesButton
+        )
+        { onPress = Just GoUp
```

Comment:
When you add destination folder to `GoUp` Msg you will have better UX for cases when browser froze and user accidentally do more clicks. In your cases if there will be multiple messages emitted from the same level you will end up in root folder. But if you will have `GoUp parentFolder` and miltiple messages will be emitted, all are producing same output (go one level up).
Also with that approach, you can simplify your `update` `case` because you can have only one message for both actions like `GoToFolder (Maybe AudienceFolderID)`.

Oh wait, I see now, it requires another argument here because `parentFolder` int this case is not `parentFolder` but `curentFolder` 🤔

Reply:
I've genteelly picked the name for the messages `GoUp` and `GoDown AudienceFolderID` because they clearly show what happens and it will be handled exactly like `GoToFolder (Maybe AudienceFolderID)`:

```elm
case msg of
    GoToFolder Nothing ->
        ...
    GoToFolder (Just id) ->
        ...
```
same with
```elm
case msg of
    GoUp ->
        ...
    GoDown id ->
        ...
```

>  for cases when browser froze and user accidentally do more clicks. In your cases if there will be multiple messages emitted from the same level you will end up in root folder. But if you will have GoUp parentFolder and miltiple messages will be emitted, all are producing same output (go one level up)

It's still JS and this part is synchronous and if this case happen it will work exactly like with `GoToFolder (Maybe AudienceFolderID)`:
- UI got frozen
- user clicks to the button [1]
- user clicks again [2]
- and again [3]
- UI is unlocked
- [1] event bubbles up and single JS stream is working under calling update (state will be changed) and then render according the new state
- UI updated and [2] event bubbles and calls the `GoToFolder` with argument from the new state from ^
- similar for [3] as ^

Result will be the same.